src/envoy/http/authnv2/filter.cc

```diff
@@ -0,0 +1,236 @@
+        (*claims)["iss"].list_value().values().Get(0).string_value() + "/" +
+            (*claims)["sub"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.audiences
+  if (claims->find("azp") != claims->end()) {
+    setKeyValue(authn_data, istio::utils::AttributeName::kRequestAuthPresenter,
+                (*claims)["azp"].list_value().values().Get(0).string_value());
+  }
+
+  // request.auth.claims
+  (*(*authn_data
+          .mutable_fields())[istio::utils::AttributeName::kRequestAuthClaims]
+        .mutable_struct_value())
+      .MergeFrom(claim_structs);
+
+  // request.auth.raw_claims
+  setKeyValue(authn_data, istio::utils::AttributeName::kRequestAuthRawClaims,
+              jwt);
+
+  return true;
+}
+
+// Returns true if the peer identity can be extracted from the underying
+// certificate used for mTLS.
+bool ProcessMtls(const Network::Connection* connection,
+                 ProtobufWkt::Struct& authn_data) {
+  std::string peer_principle;
+  if (connection->ssl() == nullptr ||
+      !connection->ssl()->peerCertificatePresented()) {
+    return false;
+  }
+  Utils::GetPrincipal(connection, true, &peer_principle);
+  if (peer_principle != "") {
+    setKeyValue(authn_data, istio::utils::AttributeName::kSourcePrincipal,
+                peer_principle);
+  }
+  return true;
+}
+
+std::string AuthnV2Filter::extractJwtFromMetadata(
+    const envoy::api::v2::core::Metadata& metadata, std::string* jwt_payload) {
+  std::string issuer_selected = "";
+  auto filter_it = metadata.filter_metadata().find(
+      Extensions::HttpFilters::HttpFilterNames::get().JwtAuthn);
+  if (filter_it == metadata.filter_metadata().end()) {
+    return "";
+  }
+  const ::google::protobuf::Struct& jwt_metadata = filter_it->second;
+  // Iterate over the envoy.jwt metadata, which is indexed by the issuer.
+  // For multiple JWT, we only select on of them, the first one lexically
+  // sorted.
+  for (const auto& entry : jwt_metadata.fields()) {
+    const std::string& issuer = entry.first;
+    if (issuer_selected == "" || issuer_selected.compare(issuer) > 0) {
+      issuer_selected = issuer;
+    }
+  }
+  if (issuer_selected != "") {
+    const auto& jwt_entry = jwt_metadata.fields().find(issuer_selected);
+    Protobuf::util::MessageToJsonString(jwt_entry->second.struct_value(),
+                                        jwt_payload);
+  }
+  return issuer_selected;
+}
+
+FilterHeadersStatus AuthnV2Filter::decodeHeaders(HeaderMap&, bool) {
+  ENVOY_LOG(debug, "AuthnV2Filter::decodeHeaders start\n");
+  auto& metadata = decoder_callbacks_->streamInfo().dynamicMetadata();
+  auto& authn_data =
+      (*metadata.mutable_filter_metadata())[Utils::IstioFilterName::kAuthnV2];
+
+  // Always try to get principal and set to output if available.
+  ProcessMtls(decoder_callbacks_->connection(), authn_data);
+  std::string jwt_payload = "";
+  std::string issuer = extractJwtFromMetadata(metadata, &jwt_payload);
```

Comment:
`issuer` is used for logging purpose only. It may be cleaner to move the logging into the function and return `void`.